src/Microsoft.Diagnostics.Monitoring.EventPipe/DistributedTraces/DistributedTracesPipeline.cs

```diff
@@ -0,0 +1,150 @@
+                await ValidateEventSourceVersion().ConfigureAwait(false);
+            }
+
+            await base.OnRun(token).ConfigureAwait(false);
+        }
+
+        private async Task ValidateEventSourceVersion()
+        {
+            int majorVersion = 0;
+
+            using CancellationTokenSource cancellationTokenSource = new();
+
+            DiagnosticsEventPipeProcessor processor = new(
+                new ActivitySourceConfiguration(1D, activitySourceNames: null),
+                async (EventPipeEventSource eventSource, Func<Task> stopSessionAsync, CancellationToken token) => {
+                    eventSource.Dynamic.All += traceEvent => {
+                        try
+                        {
+                            if ("Version".Equals(traceEvent.EventName))
+                            {
+                                majorVersion = (int)traceEvent.PayloadValue(0);
+                            }
+
+                            if (!cancellationTokenSource.IsCancellationRequested)
+                            {
+                                // Note: Version should be the first message
+                                // written so cancel once we have received a
+                                // message.
+                                cancellationTokenSource.Cancel();
+                            }
+                        }
+                        catch (Exception)
+                        {
+                        }
+                    };
+
+                    using EventTaskSource<Action> sourceCompletedTaskSource = new(
+                        taskComplete => taskComplete,
+                        handler => eventSource.Completed += handler,
+                        handler => eventSource.Completed -= handler,
+                        token);
+
+                    await sourceCompletedTaskSource.Task.ConfigureAwait(false);
+                });
+
+            try
+            {
+                await processor.Process(Client, TimeSpan.FromSeconds(10), resumeRuntime: false, token: cancellationTokenSource.Token).ConfigureAwait(false);
+            }
+            catch (OperationCanceledException)
+            {
+            }
+
+            await processor.DisposeAsync().ConfigureAwait(false);
+
+            if (majorVersion < 9)
+            {
+                throw new PipelineException("Sampling ratio can only be set when listening to processes running System.Diagnostics.DiagnosticSource 9 or greater");
+            }
+        }
+
+        protected override async Task OnEventSourceAvailable(EventPipeEventSource eventSource, Func<Task> stopSessionAsync, CancellationToken token)
+        {
+            await ExecuteActivityLoggerActionAsync((logger) => logger.PipelineStarted(token)).ConfigureAwait(false);
+
+            eventSource.Dynamic.All += traceEvent => {
+                try
+                {
+                    if (traceEvent.TryGetActivityPayload(out ActivityPayload activity))
```

Comment:
Couldn't we move the validation code here and get away with just one session? We are not validating prior to running anyway.